fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabsTray.kt

```diff
@@ -145,60 +146,77 @@ fun TabsTray(
+                                        onHeaderClick = onInactiveTabsHeaderClick,
+                                        onDeleteAllButtonClick = onDeleteAllInactiveTabsClick,
+                                        onAutoCloseDismissClick = {
+                                            onInactiveTabAutoCloseDialogCloseButtonClick()
+                                            showAutoCloseDialog = !showAutoCloseDialog
+                                        },
+                                        onEnableAutoCloseClick = {
+                                            onEnableInactiveTabAutoCloseClick()
+                                            showAutoCloseDialog = !showAutoCloseDialog
+                                        },
+                                        onTabClick = onInactiveTabClick,
+                                        onTabCloseClick = onInactiveTabClose,
+                                    )
+                                }
                             }
-                        }
 
-                        if (showInactiveTabsAutoCloseDialog) {
-                            onInactiveTabsAutoCloseDialogShown()
-                        }
+                            if (showInactiveTabsAutoCloseDialog) {
+                                onInactiveTabsAutoCloseDialogShown()
+                            }
 
-                        TabLayout(
-                            tabs = normalTabs,
-                            displayTabsInGrid = displayTabsInGrid,
-                            selectedTabId = selectedTabId,
-                            selectionMode = multiselectMode,
-                            onTabClose = onTabClose,
-                            onTabMediaClick = onTabMediaClick,
-                            onTabClick = handleTabClick,
-                            onTabLongClick = onTabLongClick,
-                            header = optionalInactiveTabsHeader,
-                        )
+                            TabLayout(
+                                tabs = normalTabs,
+                                displayTabsInGrid = displayTabsInGrid,
+                                selectedTabId = selectedTabId,
+                                selectionMode = multiselectMode,
+                                onTabClose = onTabClose,
+                                onTabMediaClick = onTabMediaClick,
+                                onTabClick = handleTabClick,
+                                onTabLongClick = onTabLongClick,
+                                header = optionalInactiveTabsHeader,
+                            )
+                        } else {
+                            Text(
+                                text = "Empty state",
+                                color = FirefoxTheme.colors.textPrimary,
+                                style = FirefoxTheme.typography.body1,
+                            )
```

Comment:
I had to add a placeholder empty state UI before the official ticket because of the `TabLayout` getting composed when there's no tabs, breaking the initial scroll because of how the tab lists are populated in `TabsTrayStore`

Reply:
This could be done in the empty state ticket, but I think we should split these conditional cases into their own functions for readability. Something like:

```
if (normalTabs.isNotEmpty() || inactiveTabs.isNotEmpty()) {
    TabsUI()
} else {
    EmptyUI()
}
```

That, or the empty UI case should come first. My complaint is mostly that the nesting in the `TabsLayout` case makes it a little hard to parse

Reply:
I do also see the "Empty state" text flash briefly when loading the tray, but I see that that is not technically a functionality degradation vs the XML version. Still, I can't help wondering if there's a way we could distinguish a "loading" state from an "empty" state to avoid that flash of text.

Reply:
For sure. We almost kinda need a tri-state here of:
1. Empty
2. Loading
3. Populated

I'm not immediately sure how to engineer that with the store observations.